frontend/src/lib/components/Table/types.ts

```diff
@@ -0,0 +1,137 @@
+import type React from "react";
+
+// --- --- --- --- --- --- --- --- --- --- --- --- --- --- --- --- --- ---
+// - Table column configuration
+export type DataColumn<TData extends Record<string, any>, TK extends keyof TData> = {
+    _type: "data";
+    columnId: TK;
+    label: string;
+    sizeInPercent: number;
+
+    hoverText?: string;
+    /** @default true */
+    sortable?: boolean;
+
+    filter?: boolean | CustomColumnFilter<TData, TK>;
+
+    formatValue?: (value: TData[TK], entry: TData) => string;
+    formatStyle?: (value: TData[TK], entry: TData) => React.CSSProperties;
+    renderData?: (value: TData[TK], entry: TData) => React.ReactNode;
+};
+
+type DataColumnAllKeys<TData extends Record<string, any>> = {
+    [TK in keyof TData]: DataColumn<TData, TK>;
+}[Extract<keyof TData, string>];
+
+// TODO -- Future work: Allow virtual columns, not directly tied to a data-value
+// export type VirtualColumn<TData extends Record<string, any>> = {
+//     _virtual: true;
+//     columnId: string;
+//     label: string;
+//     sizeInPercent: number;
+
+//     // Requires a render function
+//     render: (datum: TData) => React.ReactNode;
+// };
+
+export type ColumnGroup<TData extends Record<string, any>> = {
+    _type: "group";
+    label: string;
+    columnId: string;
+    hoverText?: string;
+    sizeInPercent: number;
+    subColumns: TableColumn<TData>[];
+};
+
+export type TableColumn<TData extends Record<string, any>> = DataColumnAllKeys<TData> | ColumnGroup<TData>;
+//  | VirtualColumn<TData>
+
+export type TableColumns<TData extends Record<string, any>> = TableColumn<TData>[];
+
+export interface ColumnFilterImplementationProps<TFilterVal = any> {
+    value: TFilterVal | null | undefined;
+    onFilterChange: (newValue: TFilterVal | null | undefined) => void;
+}
+
+export type CustomColumnFilter<TData extends Record<string, any>, TDataKey extends keyof TData> = {
+    render?: (props: ColumnFilterImplementationProps<any>) => React.ReactNode;
+    predicate?: (
+        filterValue: any,
+        dataValue: TData[TDataKey],
+        dataDef: DataCellDef<TData, TDataKey>,
+        entry: TData,
+    ) => boolean;
+};
+
+export type PendingData = { _pending: true };
```

Comment:
whis is it with underscore suffix? In usage we have to configure a pending row by setting `_pending: true` to actually render pending rows. In my head it should then be a regular public syntax?